votor/src/consensus_pool/vote_pool.rs

```diff
@@ -1,129 +1,304 @@
+        assert_eq!(self.slot, vote.vote.slot());
+        match vote.vote {
+            Vote::Notarize(notar) => {
+                if self.skip.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.notar.entry(voter) {
+                    Entry::Occupied(e) => {
+                        // unwrap should be safe as we should only store notar type votes here
+                        if e.get().vote.block_id().unwrap() == &notar.block_id {
+                            Err(AddVoteError::Duplicate)
+                        } else {
+                            Err(AddVoteError::Slash)
+                        }
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::NotarizeFallback(nf) => {
+                if self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.notar_fallback.entry(voter) {
+                    Entry::Vacant(e) => {
+                        let mut map = BTreeMap::new();
+                        map.insert(nf.block_id, vote);
+                        e.insert(map);
+                        Ok(())
+                    }
+                    Entry::Occupied(mut e) => {
+                        let map = e.get_mut();
+                        let len = map.len();
+                        match map.entry(nf.block_id) {
+                            Entry::Vacant(map_e) => {
+                                if len == 3 {
+                                    Err(AddVoteError::Slash)
+                                } else {
+                                    map_e.insert(vote);
+                                    Ok(())
+                                }
+                            }
+                            Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                        }
+                    }
+                }
+            }
+            Vote::Skip(_) => {
+                if self.notar.contains_key(&voter) || self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.skip.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::SkipFallback(_) => {
+                if self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.skip_fallback.entry(voter) {
+                    Entry::Occupied(_) => Err(AddVoteError::Duplicate),
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
```

Comment:
There is some code duplication across cases, would be best to put them in a small function, but I can live with the copy and paste here.